invoice assistant

kevinfink96 opened the pull request:
@plentymarkets/team-order-payment 

Kanbanize: https://plentymarkets.kanbanize.com/ctrl_board/75/cards/68627/details/

kevinfink96 committed:
added form elements

kevinfink96 committed:
edited ui and created save function

kevinfink96 committed:
edited save function and added some basic description text

kevinfink96 committed:
changed wrong english description

kevinfink96 committed:
load old settings from mysql

kevinfink96 committed:
added new method getInvoiceClient

kevinfink96 committed:
removed old settings and finished AssistantDataSource

kevinfink96 committed:
fixed bug with loading old settings & removed unnecessary required=false fields

Commit:
cleanup for format
change translations
change wording

Commit:
fix wording

kevinfink96 committed:
delete old settings ui

kevinfink96 committed:
fix multilingualism page and edited user_guide

kevinfink96 committed:
user_guide_en

kevinfink96 committed:
update changelog and plugin version

DDBreidenbach commented on meta/documents/changelog_en.md:
```diff
@@ -1,5 +1,10 @@
+## 2.0.0 (2019-08-15)
+ 
+### Note 
+- The settings for the Invoice plugin have been transferred to an assistant in the **System » Assistants » Payment**
```
```suggestion
- The settings for the Invoice plugin have been transferred to an assistant in the **System » Assistants » Payment** menu.
```

DDBreidenbach reviewed (commented):
wording

kevinfink96 committed:
Update meta/documents/changelog_en.md

Co-Authored-By: Denise Breidenbach <38553920+DDBreidenbach@users.noreply.github.com>

kevinfink96 committed:
removed commented lines
added annotations

kevinfink96 committed:
Merge branch 'order/feature/assistant' of https://github.com/plentymarkets/plugin-payment-invoice into order/feature/assistant

raphael-adam commented on src/Assistants/DataSources/AssistantDataSource.php:
```diff
@@ -0,0 +1,149 @@
+        foreach ($pids as $pid) {
+            $settings = $this->settingsService->loadClientSettingsIfExist($pid, null);
+            if (count($settings)) {
+                $settings = $this->settingsService->getSettingsForPlentyId($pid, null);
```
Ich verstehe nicht was hier gemacht wird. Du lädst die Settings und wenn sie gesetzt sind, lädst du sie nochmal über eine andere Methode?

kevinfink96 replied:
Werde die Funktion etwas bearbeiten und umbenennen, weil sie eigentlich nur da war um zu überprüfen, ob es bereits Einstellungen gibt. In getSettingsForPlentyId werden die Daten nicht nur geladen, sondern auch verarbeitet und in die richtige Form gebracht.

raphael-adam commented on src/Assistants/DataSources/AssistantDataSource.php:
```diff
@@ -0,0 +1,149 @@
+            $settings = $this->settingsService->loadClientSettingsIfExist($pid, null);
+            if (count($settings)) {
+                $settings = $this->settingsService->getSettingsForPlentyId($pid, null);
+                $data[$pid]['config_name'] = $pid;
```
Du musst `$data[$pid]` zuerst initialisieren, bevor du es hier als Array verwendest. Also `$data[$pid] = []`.

raphael-adam commented on src/Assistants/DataSources/AssistantDataSource.php:
```diff
@@ -0,0 +1,149 @@
+            if (count($settings)) {
+                $settings = $this->settingsService->getSettingsForPlentyId($pid, null);
+                $data[$pid]['config_name'] = $pid;
+                $data[$pid] = $settings;
```
Hier überschreibst du wieder das `$data[$pid]['config_name'] = $pid;`

raphael-adam commented on src/Assistants/InvoiceAssistant.php:
```diff
@@ -0,0 +1,406 @@
+                                "internal_info_page" => [
+                                    "type" => 'category',
+                                    'defaultValue' => '',
+                                    'isVisible' => "info_page_toggle == true && info_page_type == 1",
```
Das `== true` macht keinen Sinn. Und brauchst du das `isVisible`, wenn oben schon das gleiche in `condition` steht?

raphael-adam commented on src/Assistants/InvoiceAssistant.php:
```diff
@@ -0,0 +1,406 @@
+
+        $webstore = $webstoreRepository->findById(0);
+
+        return $webstore->storeIdentifier;
```
Hier und bei `getWebstoreListForm` würde ich auch mit Caching arbeiten, so wie es bei `getLanguage` und bei `getCountriesListForm` schon drin ist

raphael-adam commented on src/Assistants/SettingsHandlers/InvoiceAssistantSettingsHandler.php:
```diff
@@ -0,0 +1,225 @@
+        $data = $parameter['data'];
+        $webstoreId = $data['config_name'];
+        if(!is_numeric($webstoreId) || $webstoreId <= 0){
+            $webstoreId = $this->getWebstore($webstoreId)->storeIdentifier;
```
Das macht irgendwie keinen Sinn: wenn die webstoreId nicht gültig ist, dann lädst du mit der webstoreId den Webstore?

raphael-adam commented on src/Assistants/SettingsHandlers/InvoiceAssistantSettingsHandler.php:
```diff
@@ -0,0 +1,225 @@
+    private function saveInvoiceSettings($webstoreId, $data)
+    {
+        $settings = [
+            'name' => $data['name'],
```
Ist bei den ganzen Array-Keys auf die du hier zugreifst sichergestellt, dass die immer gesetzt sind? Wenn nicht, müsstest du zuerst prüfen ob die gesetzt sind. Das geht z.B. so: `$data['name'] ?? ''`. Das ist das gleiche wie `isset($data['name']) ? $data['name'] : ''`

raphael-adam commented on src/Assistants/SettingsHandlers/InvoiceAssistantSettingsHandler.php:
```diff
@@ -0,0 +1,225 @@
+        $settingsService = pluginApp(SettingsService::class);
+        $getSettings = $settingsService->loadClientSettingsIfExist($webstoreId,$this->getLanguage());
+        if(!count($getSettings)){
+            $settingsService->updateClient($webstoreId);
```
Das ist irgendwie eine komische Logik. Warum kümmert sich nicht `saveSettings` darum?

kevinfink96 replied:
Am Anfang wollte ich die saveSettings Funktion nicht ändern, damit die alte Logik weiterhin funktioniert, aber das war unnötig, weil die alte Settings UI entfernt wird.

raphael-adam commented on src/Services/SettingsService.php:
```diff
@@ -367,10 +413,41 @@ private function loadClientSettings($plentyId, $lang)
+        {
+            $query->where('lang', '=', $lang);
+        }
+        $query->orWhere('lang',   '=', '')->where('plentyId', '=', $plentyId);
```
Das `orWhere` macht doch nur Sinn, wenn `$lang` gesetzt ist. Und hier ist zum zweiten Mal die Einschränkung auf die plenty ID.

raphael-adam commented on src/Services/SettingsService.php:
```diff
@@ -367,10 +413,41 @@ private function loadClientSettings($plentyId, $lang)
+        /** @var Settings[] $clientSettings */
+        $clientSettings = $query->get();
+
+        if( !count($clientSettings) > 0)
```
Das macht gar keinen Sinn. Das `!` gilt nur für das `count`, nicht den ganzen Ausdruck. Du kannst es einfach so schreiben: `if (!count($clientSettings))`. 0 evaluiert zu false, alles darüber zu true.

raphael-adam reviewed (changes requested):
Es muss auf jeden fall nochmal jemand drüber schauen, der sich mit dem Plugin und mit Assistenten auskennt.
Allgemein ist auch dein Codestyle nicht richtig und du hast oft unnötigte Leerzeichen und Zeilenumbrüche drin.

kevinfink96 committed:
mainWebstore and webstoreValues to class variable
renamed laodClientSettingsIfExist to clientSettingsExist and optimized code
default values for values on saving

kevinfink96 committed:
changed if equals true condition

kevinfink96 committed:
load plentyId if webstore is invalid

kevinfink96 committed:
add .iml files to gitignore

kevinfink96 committed:
Delete Invoice.iml

MaselP reviewed (changes requested):
Zusatzinformationen und Logo sind falsch bzw werden nicht aus der alten konfig übernommen
und oben bei verfügbarkeit ist das "Zahlungsart erlauben" ein bisschen irreführend
mach am besten "Zahlungsart für Gäste erlauben" oder sowas

![Bildschirmfoto 2019-09-03 um 12 21 55](https://user-images.githubusercontent.com/10023830/64165659-caae5a00-ce45-11e9-8bfe-29c3625169e4.png)

MaselP commented:
es gab in der alten konfig auch noch eine "Beschreibung" oder "description" 
Wird die nicht mehr benötigt?

kevinfink96 commented:
> es gab in der alten konfig auch noch eine "Beschreibung" oder "description"
> Wird die nicht mehr benötigt?

gab name und beschreibung, was beim lastschrift plugin auch rausgeflogen ist

kevinfink96 committed:
loading missing data from old settings
checks the container links if icon in footer is enabled

kevinfink96 committed:
Merge branch 'order/feature/assistant' of https://github.com/plentymarkets/plugin-payment-invoice into order/feature/assistant

kevinfink96 committed:
checks the container links if icon in footer is enabled

kevinfink96 committed:
added missing translation

kevinfink96 committed:
selectbox default select fix

kevinfink96 committed:
fix webstore is null

kevinfink96 committed:
check if pluginSetId is null

kevinfink96 committed:
fix small bug with info-link and logo toggle

kevinfink96 committed:
userguide: changed systemtree

kevinfink96 committed:
user guide: changed systemtree

kevinfink96 committed:
only set container key if bank option is set

kevinfink96 committed:
text: change system tree

kevinfink96 committed:
removed whitespace and deleted duplicate translation key for the english texts

kevinfink96 committed:
fix for bank container layout connection

raphael-adam commented on src/Assistants/DataSources/AssistantDataSource.php:
```diff
@@ -0,0 +1,178 @@
+            $settingsExist = $this->settingsService->clientSettingsExist($pid, null);
+            if ($settingsExist) {
+                $settings = $this->settingsService->getSettingsForPlentyId($pid, null);
+                $data[$pid] = [];
```
Das macht so keinen Sinn, weil du es in der nächsten Zeile wieder überschreibst.

raphael-adam commented on src/Assistants/InvoiceAssistant.php:
```diff
@@ -0,0 +1,423 @@
+            $webstores = $this->webstoreRepository->loadAll();
+            /** @var Webstore $webstore */
+            foreach ($webstores as $webstore) {
+                $this->webstoreValues[] = [
```
Du benutzt `$this->webstoreValues` als Array ohne es vorher als Array initialisiert zu haben.

raphael-adam commented on src/Services/SettingsService.php:
```diff
@@ -367,10 +412,34 @@ private function loadClientSettings($plentyId, $lang)
+        {
+            $query->where('lang', '=', $lang);
+        }
+        $query->orWhere('lang',   '=', '')->where('plentyId', '=', $plentyId);
```
Das hier hast du doch gar nicht angepasst.

raphael-adam commented on src/Services/SettingsService.php:
```diff
@@ -367,10 +412,34 @@ private function loadClientSettings($plentyId, $lang)
+
+        /** @var Settings[] $clientSettings */
+        $clientSettings = $query->get();
+        return count($clientSettings) > 0;
```
Du kannst auch direkt `$query->exists()` aufrufen, das ist performanter.

kevinfink96 replied:
Geht nicht im Plugin.

kevinfink96 committed:
requested changes

kevinfink96 committed:
refactor: check if client settings exist

kevinfink96 committed:
remove parameter lang from clientSettingsExist function

kevinfink96 committed:
add language to settings

raphael-adam reviewed (approved):
Die von mir angemerkten Stellen müssten jetzt passen. Aber im Allgemeinen muss noch einer mit mehr Assistent-Sachverstand drüber schauen.

kevinfink96 committed:
naming

Commit:
payment description can now be changed via CMS => multilingualism

Commit:
changelog update

MaselP reviewed (approved):
💎

kevinfink96 committed:
datasource: allow payment for guests checkbox fix

kevinfink96 committed:
empty description text as default

kevinfink96 committed:
refactor allowInvoiceForGuest

kevinfink96 committed:
fix: standard logo won't show

kevinfink96 committed:
updated system tree in the changelog

kevinfink96 committed:
save settings global and not per language
disabled checkbox for logo in the footer by default

kevinfink96 committed:
update changelog date

kevinfink96 committed:
moved payment method name to CMS -> multilingualism

kevinfink96 committed:
changelog wording
